### ISSUE TITLE
[Dev Asks] Anybody up for a Station concept?

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#6065]
So I was watching 2001 with my son and I had a silly idea while seeing this beauty:

![image](https://github.com/user-attachments/assets/5191466f-b9e7-4f19-bdd7-f5443e25c7c9)

Anybody wanting to give a shot a making an inspired station?

It would have to be modular:
 - One module for the center
 - One module for the column going from the center to the ring
 - Finally a module for a section of the outer ring. This section will be flat and dock to a neighboring section at a slight angle. This is already possible with the current station framework (I still need to integrate Halo ring and warp gate). This module could come in several variant, some spawning kiosks, some with other stuff. It would also need to connect to a vertical section. We would spawn 16+ or so modules, it would look rounded enough.

There is plenty of other things to decide and discuss (especially on the center section, and how to navigate to the outer ring) but I want to see if anybody would be up for this. This is not super important for gameplay perspective, but it would add a nice variety to the existing stations. I would either add an option to have the station perpetually rotating, or give an infinitely moving rotator.
Building-wise it might be tricky to nail at first, but the proof of concept was already done with the warp gate.

![image](https://github.com/user-attachments/assets/28007b2f-06ae-4117-9ffd-d732fa892f24)

Cheers

**Comment by ProPeach** [comment]
This would be very cool to see! I've seen a lot of people want to try a station like this but the lack of infinitely rotating rotor has put them off, great to hear you're considering adding one.

Couple questions - 

What orientation would you want the gravity? The cheap answer would be in line with the axis of rotation so there is no awkward transitions, but of course the whole point of a station like this is to have radial gravity generatied by the spin. This would mean elevators going up the columns and a gravity transition in the middle hub. Both are totally possible of course, but the choice makes a big difference in how it's made.

And of course, how big? How large a ship/how many do you want docking for in the middle, and how large a radius do you want for the ring?

**Comment by tsunamayo** [comment]
@CoasterBlock Those are impressive habitat indeed! So I wanted to migrate the current ring to a brick based system, using the new terrain tech (habitat is still old one, you cant mine it). It will take a bit of time as there is ton of thing to migrate, but I would give a (creative only) habitat computer to generate terrain that fit your design. I think you would like that update ^^
Habitat would be for very large ring, the station I have in mind is much smaller, I would say maybe 200 m max I need to do a few measurement and calculation.

@ProPeach so the best practice is to have a teleport station near the kiosks, as it might be fun to do a bit of walk at first but after 50x time the player might want to move faster. 
For the gravity on the ring, I think using artificial gravity will make sense, even if the thing is rotating a bit. Yes we would use cheated life support like for the other stations (will also make the game run faster)

So yes there are things open to discuss:
1) Do we put a hangar in the middle? If yes do we have the hanger static or does it rotate? I could have an option for the gravity field that would also rotate nearby entity, so landing could be easy, or we could let the auto-pilot figuring it out. I think it would just for show anyway, as it cant be too big. Maybe something close the tutorial module by Surge.

2) vertical column moving strategy: elevator? 0 g corridor? honestly I was considering having a gravity lift: a down section with an to outside gravity, and an up section with an inward gravity. The player would essentially fall. It feel more scifi and easier to build.

3) the size, let me do a few tests

4) module to be spawned: we need a kiosk one. Other ideas: habitat (with some mini appartement), leisure / sightseeing with more windows. Tech / maintenance section. Botanical garden.

**Comment by SovietSponge** [comment]
Little apartment modules sound really nice.
As for the hangar position I think it'd be really cool to have some variety in it, some stations having a central hangar and others having them on the rings while the kiosks n such could be in the center.

**Comment by Uncle-Ulty** [comment]
Hi,

I'm interested in this. I started designing a station 2 years ago, but never completed it.

There are a lot of parameters to set before we start building. 
For easy coding purposes, the static hangar in the middle would be the best option. However, a 200 m diameter station would be too small to fit a center hangar. 
The design would look too crowded in the middle, without the nice sleek profile of the columns connecting the ring to the center of the concept design.
 Besides, a 200 m diameter toroidal station would be too narrow to park a 100 m ship...
 
 So, I believe we must start from the center module. We need some guidelines about how many landing pads are necessary in the middle. Depending on how many are required, the station would be too big (I believe that's the reason you want to put the hangars on the ring, to have more available space for them!
 
 Also, how many bricks are acceptable for each part? center, column and ring?
 
 That's a nice challenge and I accept it! :D
 
 cheers!

**Comment by CoasterBlock** [comment]
@tsunamayo That terrain computer does sound very much up my alley lol! Cant wait. 

As for your other points, I have done some experiments using my designs already, and the main issue I encounter is gravity + rotation. Static rings with radial gravity work fine, but when the entity is moving it could cause issues. So perhaps the player should move with the entity as long as there is gravity and the player is in the bounding box of the CE or Station. That would mitigate some buggyness. 

I had success by just making a rail or piston elevator for the main connector beams from the hub to the ring. However, a gravity tunnel also works where a gravity gen moves the player. Elevators are the classic solution, gravity the scifi solution. 

The hub is better stationary I think. If the ring is supposed to spin at one point, then keeping it in the center would be easier and more practical. Both for building and for the landing-AI. 

a space-station with a ring id go for 200m - 1000m maximum. Anything larger and it becomes to long to tranverse unless someone figures out trains. 

As for air-tightness, perhaps a "ring-computer" could do some of the work? If a ship or station is designed using a ring-computer life-support and air would work by just using entity bounds to speed up the calculation, and allow for an easier and more accesible ring building experience. That way rings could be airtight, even if there is small gaps between each segment?

**Comment by Uncle-Ulty** [comment]
For stations, it would be nice if we could "build" trees and new plants for atriums, and maybe small farms.

**Comment by CoasterBlock** [comment]
@tsunamayo I dug out another design I made a few weeks ago, 283m, a prototype I wanted to turn into a ship. Here it is next to to the V27 Hotwing included in the game. 

![20241026193324_1](https://github.com/user-attachments/assets/de3afc8c-53ca-4ba1-8667-2f763ad7a34a)

I agree that 200m may be a little to small. I suggest 500m minimum as a diamter to make sure there even is enough space for everything. Including space to dock a larger ship.

**Comment by tsunamayo** [comment]
Sorry let me get back to you all asap!

**Comment by tsunamayo** [comment]
@Uncle-Ulty Hi, glad to see a talented builder like you up to the challenge!
So I think a hanger that can house a 100m ship is way too big. I am thinking about small hangar only.
I dont plan to put hangars on the outer ring: we either put a hangar in the middle or we put no hangar. It would be really complicated to land on the outside as it rotates.

The plan is to nail the size and spawn first, sort out all questions and issue we get, then do the proper build after. Otherwise there is a high chance of having to throw away a cool looking section that cant be used for X reason.
The brick count should be really low, I think under 1000s brick for a module would be great.

I did a few tests for the size, let/s start with something in the 400m-500m diameter ballpark. The middle hangar should have a 60-20m opening (roughly).
Lets use my idea of the g-elevator, then lets have the entire station rotate. I will do some test to see how complicated it is to land. 
For now dont spend much time on making good looking stuff, only have module of the right shape with the right docking params.

**Comment by tsunamayo** [comment]
@CoasterBlock dont worry for gravity and air, we will use artificial gen and entity wide air pressure. Thanks

**Comment by CoasterBlock** [comment]
@tsunamayo Thanks for that info, and thanks for providing some information about the design of the ring. Will the infinite-rotation happen using a normal rotor? 400-500m is an excellent size to work with, as each segment of the ring can be made using very few bricks. Also just so I understand before building: Each segment is individual, but will spawn together into a ring right? And the 0g tower and Hub are also a stand alone element that generates? Or is it one segment with a 0g tunnel and hangar attached?

Thanks in advance!

**Comment by Uncle-Ulty** [comment]
@tsunamayo Ok! I gonna build a functional structure first to check if it's viable. 1000 bricks per module? It's quite a challenge since it requires an interior. Maybe to use more bricks in some and fewer in others in order to average them in 1000 brick/module.

**Comment by Uncle-Ulty** [comment]
@tsunamayo 

My concept for the new station:

![image](https://github.com/user-attachments/assets/9e145cbe-a57f-4639-8a52-d4a5b108c151)

- In this concept, the mid section has radial orientation gravity, with 4 hangars, each one with a different gravity direction. 
- A rotating ring connects all hangars, the four small and the big ones (I'm using the infinite rotation trick). 
- The ring has 4 shafts for gravity elevator, each shaft has one tube to go out and another tube to go in.
- Each shaft has 2 docking ports to spawn each column that connects the mid part to the outer ring part. These two docking ports are already connected to a pipe and a conveyor

I need to build the columns, the outer ring, and improve mid part exterior aesthetic

I hope you like it!
[TEC - Ring station concept 09.zip](https://github.com/user-attachments/files/17592072/TEC.-.Ring.station.concept.09.zip)

https://youtu.be/CMamwis2CYM

**Comment by tsunamayo** [comment]
@Uncle-Ulty That is very impressive, I cant get the rotation started though. I can see it working!
I was thinking, if you do 4 hangar which each their gravity, it might be even more impressive for the player to have the 4 hangars merged into a big one, each which its local gravity on each side. So that when you look up you see the other side with potentially other ships parked. All for that scifi woa factor, which is kinda the theme of that station ;)
Proportion wise will have to make sure it works, so lets wait and see with the column and outer ring. It is seems big already, maybe too long, it the worst case we will shorten or get rid of the very large hangar.

The middle section that rotate is really really nice, makes me want to make a mini rotating station too haha, maybe an idea for another time.
https://www.youtube.com/watch?v=1wJQ5UrAsIY

Good work, cheers!

**Comment by Uncle-Ulty** [comment]
@tsunamayo That's a great idea! All of them in the same space, but different orientations. 

About the size, I did some math to keep under the 500 m size.

I believe the big ships must have airlocks for docking.

**Comment by Uncle-Ulty** [comment]
@tsunamayo , Is the station generator currently working? I can't generate my new station atm, but I can generate my old warp gate...

also, it would be good to have some help tutorial concerning station generation

**Comment by tsunamayo** [comment]
@Uncle-Ulty Yes it should work. Send me your setup and I will see what is wrong.

**Comment by Tackietacktack** [comment]
I dont want to make a great discussion,

My question is, must this station be a simple build(low brick count, low ce count)? 

I found out, that local build (not controlled) don't bad for game pervormance, which have in a modul massiv light or CEs.

Example is the ship from #6089 :

I have spawn 20 piece from this ship
It have 5000 CEs
Round about 1000 lights
And i can controll one ship from it with no lags

**Comment by Uncle-Ulty** [comment]
@tsunamayo 

I'm making progress with the ring station.

- new trusses on columns to make it look sturdier
- updated habitat to look as like a mixture of plaza park and food court

![Image](https://github.com/user-attachments/assets/890539ba-d084-4149-b556-629cacd71516)
Ring station with 16 sections on ring

![Image](https://github.com/user-attachments/assets/d272a4e0-60d0-43d4-a3df-86699791bf7f)
habitat section ( 5.000 bricks only, far inferior than surge and lord comissair's hubs)

Trees and vegetation makes it look even prettier

Every tree patch is placed as a decoration kiosk, therefore it doesn't add brick count on the ring section,  since these trees are props that use each 1000 bricks. That's why we need new natural bricks with trees, brushes, stones, grass and other natural materials.

**Comment by CoasterBlock** [comment]
@Uncle-Ulty That Looks fantastic! Well done!
